### DISCUSSION TITLE
Expose current HBAR price as "oracle data"

### DISCUSSION THREAD

**Discussion by Cooper-Kunz** [hiero-ledger/hiero-improvement-proposals#811] [Ideas]
Currently, Hedera has a [fee schedule](https://github.com/hashgraph/hedera-services/blob/ccb04ea45f3e6d1a8574128c75813eef9904b8f6/hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/ExchangeRateManager.java#L47) that is dependent on knowing the current USD price of HBAR.. Or perhaps preferred, I would assume the total volume weighted average price (TVWAP) over a specific time period.

"All fees in Hedera are based on the exchange rate between HBAR and USD. Fees are paid in HBAR, but based on the current USD price of the HBAR. The "ERT", exchange rate tool, is responsible for tracking the exchange rate of various exchanges, and updating the {ExchangeRateSet} on a periodic basis. Currently, this is in a special file, but could be from any other source. The encoded {Bytes} are passed to the {update(Bytes)} method. This MUST be done on the same thread that this manager is used by -- the manager is not threadsafe."

This means that the network is consuming prices from centralized exchanges, processing that data, and acting on it via a file update transaction... Which means the network has _some practical, but potentially limited_ native support for HBAR price oracles! However,  the network (& therefore council) is choosing not to expose the Exchange Rate Tool (ERT) info to 3rd party devs... 

I'm opening up a thread for the following reasons -

1. I constantly hear that you cannot build X on Hedera due to not having oracles (e.g. lending protocols..)
2. Hedera _technically_ uses an oracle to calculate it's own fee schedule
3. Why are we not exposing that information to 3rd party developers?

> Note: why Chainlink doesn't deploy to Hedera is outside of the scope of this discussion, but a _really_ good question! 

I suggest that instead of the current 1 hour cadence at which Hedera updates their exchange rate, Hedera publishes this data on a regular interval, e.g every ~13 seconds (mapping to Ethereum's block times and therefore Defi protocol expectations) to the smart contract service, file service, and consensus service, for developers to consume within their applications. I expect many will want to be able to "pull" this information on demand, but that is likely more engineering work.

Unlike many of my other governance proposals / HIPs I do not _currently_ have any intentions to use this one, and so may not be the best stakeholder to explicitly define requirements, goals, or otherwise. It is likely I will not see this idea through to implementation, but I do want to pose the question.... why are we not exposing this information (i.e. the current price of HBAR) natively, when the network already depends on having that information?

**Comment by steven-sheehy** [comment]
Note mirror nodes expose the exchange rate file via its `/api/v1/network/exchangerate` [REST API](https://mainnet-public.mirrornode.hedera.com/api/v1/docs/#/network/getNetworkExchangeRate). This `0.0.112` Hedera network file is also available via HAPI `FileGetContentsQuery` and via smart contracts via the exchange rate [precompile](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/exchange-rate-precompile/IExchangeRate.sol). Since it's in the record stream, users could also consume that and expose it however they like. Are these not sufficient sources?

**Reply by Cooper-Kunz** [reply]
That is quite possible -- perhaps my discussion post is misguided & outdated! Thanks for the quick response, Steven. 

Didn't know it was exposed in the SC precompile based on recent conversations I had.  Can you clarify how this gets surfaced to the smart contracts, the frequency it gets updated, and any docs links that may be relevant here? 

Otherwise, will look into it a bit more and get back to you on my thoughts! 😄

**Reply by Cooper-Kunz** [reply]
Note that I updated this to reflect that this may already be implemented (based on your responses above, etc.)

**Reply by Cooper-Kunz** [reply]
It does explicitly say  "This rate is a weighted median of the the recent" HBAR-USD exchange, rate on major exchanges, but should _not_ be treated as a live price  oracle!" sooo I don't believe it satisifes my requirements? curious why it says that..

**Comment by steven-sheehy** [comment]
The exchange rate tool has a [README](https://github.com/hashgraph/hedera-exchange-rate-tool) that explains how it works. It runs once an hour and submits a FileUpdate/FileAppend HAPI transaction with the [ExchangeRate](https://github.com/hashgraph/hedera-protobufs/blob/main/services/exchange_rate.proto) proto as its contents. This file is stored in state and the precompile is served from that. Those file transactions are pulled by mirror node and used as its source.

**Reply by Cooper-Kunz** [reply]
Yes I am aware of the 1 hr, thanks for the other info. So safe to assume the reason it says it shouldn't be used as a live price oracle is cause it's updated 1/hr? My proposal explicitly suggests one thing -- that we instead get this info every 13 seconds, and upload to different parts of the network, so that you _can_ use it as a live price oracle. That's literally the entire HIP idea 😄

**Reply by Cooper-Kunz** [reply]
Somewhat related but is this accurate/up to date on which exchanges you're pulling from (currently in the readme)? 

<img width="618" alt="Screenshot 2023-09-15 at 7 06 56 PM" src="https://github.com/hashgraph/hedera-improvement-proposal/assets/11776938/cafb887d-2d46-45f1-b8f0-7fed8e7f088c">

If so, any reason coinbase isn't included versus some of these "smaller" exchanges?

<img width="826" alt="Screenshot 2023-09-15 at 7 08 44 PM" src="https://github.com/hashgraph/hedera-improvement-proposal/assets/11776938/a832ca6f-67e6-4668-82bc-876fc063ee8b">

(img from coinmarketcap right now)

**Comment by steven-sheehy** [comment]
The ERT and exchange rate file are only intended for use by consensus nodes to ensure hbar fees are fixed to USD. Since consensus nodes only need or use the hourly value it doesn't make sense to update the API to change it produce it every 13s. It's not intended to match exactly what the markets are doing. In fact, the exchange rate value can be smoothed out during periods of high volatility so as to not break apps. The exchange rate APIs are useful for questions like "How much did my transaction cost in USD at time X in the past". I'm sure there's a whole swath of legal reasons why Hedera can't offer a live pricing oracle, but I am not a lawyer.

**Reply by Cooper-Kunz** [reply]
> I'm sure there's a whole swath of legal reasons why Hedera can't offer a live pricing oracle, but I am not a lawyer.

I am *very* specifically asking why this is the case, given how easy it is to technically implement with the ERT, etc.

**Reply by Cooper-Kunz** [reply]
Or even relative to the other options you have outlined for getting this information on a slower, less regular pace, I'm curious why that is okay but the other is not...? Currently the information I have received here seems to be that hedera supports price oracles! But only if you're okay with 1hr of lag, and therefore a lot of arbitrage, but not realtime, which legitimately & seriously could hurt users. But not realtime.. which is weird! right?

**Comment by web3-nomad** [comment]
the challenge isn't obtaining the price. there are various places which offer APIs to get this, and anyone could just call these APIs and push that at regular intervals into the EVM (or just store it in a file somewhere if you aren't using it in the EVM).

but if you want to provide it in the EVM for other people to use there are three challenges/issues which immediately come to mind.

- how do you provide the feed in a way that people are used too?. On EVMs this is chainlink compatible for most people. (if you're writing a new one, then you don't really care I guess).

- how do you 'trust' the person putting it on the chain, if it's your app and you are a bank/have a reputable legal entity behind you, then you can trust yourself I guess or the banks reputation, but for others they need to trust the provider isn't doing something fishy. (watch the sting for an example of what people in the middle of information could potentially do)
 
- what happens when the provider gets it wrong? (which does happen, and some exchanges get it wrong regularly for it to be an issue)

so while Hedera uses price information to set fees, the impact of getting it wrong is quite low, and there aren't many people who stand to gain if it gets subverted.

General price oracles are used a common attack vector in most defi applications, so they need to be quite solid to be used by 3rd parties. 

and again.. 
If you need one for your app, binance, coinbase, and apilayer provide feeds.

**Reply by Cooper-Kunz** [reply]
I'm confused to be quite honest... 

1. You could target the chainlink interfaces if necessary (?) but I don't believe that is as critical as having access to the data reliably. Alternatively they could come up with an improved interface! I'm sure we've all learned alot since chainlink defined theirs. 
2. The currently ERT tool is used _in the Hedera network_ & if it's secure enough for them, why is it not secure enough for others? It's quite possible to imagine there being a bug and it 10-100x-ing fees which would cause users that are not adequately setting maxTxFees to lose money, etc. & by definition of using hedera you're trusting their governing council as they're the only ones running the infra, and can update it whenever they want. So it doesn't seem like a new trust assumption to me. 
3. Again, hedera is currently using this tool for their own fee pricing mechanisms, and takes into account multiple exchanges (but arguably questionable ones as I asked [here](https://github.com/hashgraph/hedera-improvement-proposal/discussions/811#discussioncomment-7018993)). You could ask this about any oracle - what if they're wrong? Should we just not build anything that wants real world data or oracles? 

Based on the responses here + those i've received elsewhere perhaps easiest for everyone to just close this thread. 😅
